{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D0_10X.bam...\n",
      "Running: /Volumes/BioinforSSD/PROGRAMS_AND_ENVS/Programs/qualimap_v2.3/qualimap bamqc -bam /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D0_10X.bam -outdir quality_check_results/D0_10X.bam_qualimap --java-mem-size=4G\n",
      "Java memory size is set to 4G\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.3\n",
      "Built on 2023-05-19 16:57\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 35\n",
      "Max memory (Mb): 4294\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 483\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 483 windows...\n",
      "Processed 100 out of 483 windows...\n",
      "Processed 150 out of 483 windows...\n",
      "Processed 200 out of 483 windows...\n",
      "Processed 250 out of 483 windows...\n",
      "Processed 300 out of 483 windows...\n",
      "Processed 350 out of 483 windows...\n",
      "Processed 400 out of 483 windows...\n",
      "Processed 450 out of 483 windows...\n",
      "Total processed windows:483\n",
      "Number of reads: 326291759\n",
      "Number of valid reads: 298044078\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 298044078\n",
      "Num mapped first of pair: 0\n",
      "Num mapped second of pair: 0\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 713\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 203809802478\n",
      "referenceSize: 3101804739\n",
      "numberOfSequencedBases: 29215645542\n",
      "numberOfAs: 8017611837\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 720\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "Finished\n",
      "Processing /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D2_10X.bam...\n",
      "Running: /Volumes/BioinforSSD/PROGRAMS_AND_ENVS/Programs/qualimap_v2.3/qualimap bamqc -bam /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D2_10X.bam -outdir quality_check_results/D2_10X.bam_qualimap --java-mem-size=4G\n",
      "Java memory size is set to 4G\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.3\n",
      "Built on 2023-05-19 16:57\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 35\n",
      "Max memory (Mb): 4294\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 483\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 483 windows...\n",
      "Processed 100 out of 483 windows...\n",
      "Processed 150 out of 483 windows...\n",
      "Processed 200 out of 483 windows...\n",
      "Processed 250 out of 483 windows...\n",
      "Processed 300 out of 483 windows...\n",
      "Processed 350 out of 483 windows...\n",
      "Processed 400 out of 483 windows...\n",
      "Processed 450 out of 483 windows...\n",
      "Total processed windows:483\n",
      "Number of reads: 328166791\n",
      "Number of valid reads: 295054047\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 295054047\n",
      "Num mapped first of pair: 0\n",
      "Num mapped second of pair: 0\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 714\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 201527471657\n",
      "referenceSize: 3101804739\n",
      "numberOfSequencedBases: 28895326240\n",
      "numberOfAs: 7968089507\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 720\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "Finished\n",
      "Processing /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D8_10X.bam...\n",
      "Running: /Volumes/BioinforSSD/PROGRAMS_AND_ENVS/Programs/qualimap_v2.3/qualimap bamqc -bam /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D8_10X.bam -outdir quality_check_results/D8_10X.bam_qualimap --java-mem-size=4G\n",
      "Java memory size is set to 4G\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.3\n",
      "Built on 2023-05-19 16:57\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 35\n",
      "Max memory (Mb): 4294\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 483\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 483 windows...\n",
      "Processed 100 out of 483 windows...\n",
      "Processed 150 out of 483 windows...\n",
      "Processed 200 out of 483 windows...\n",
      "Processed 250 out of 483 windows...\n",
      "Processed 300 out of 483 windows...\n",
      "Processed 350 out of 483 windows...\n",
      "Processed 400 out of 483 windows...\n",
      "Processed 450 out of 483 windows...\n",
      "Total processed windows:483\n",
      "Number of reads: 327366657\n",
      "Number of valid reads: 295318090\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 295318090\n",
      "Num mapped first of pair: 0\n",
      "Num mapped second of pair: 0\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 877\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 274043684440\n",
      "referenceSize: 3101804739\n",
      "numberOfSequencedBases: 28858359954\n",
      "numberOfAs: 8199288915\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 883\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "Finished\n",
      "Processing /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D12_10X.bam...\n",
      "Running: /Volumes/BioinforSSD/PROGRAMS_AND_ENVS/Programs/qualimap_v2.3/qualimap bamqc -bam /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D12_10X.bam -outdir quality_check_results/D12_10X.bam_qualimap --java-mem-size=4G\n",
      "Java memory size is set to 4G\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.3\n",
      "Built on 2023-05-19 16:57\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 35\n",
      "Max memory (Mb): 4294\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 483\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 483 windows...\n",
      "Processed 100 out of 483 windows...\n",
      "Processed 150 out of 483 windows...\n",
      "Processed 200 out of 483 windows...\n",
      "Processed 250 out of 483 windows...\n",
      "Processed 300 out of 483 windows...\n",
      "Processed 350 out of 483 windows...\n",
      "Processed 400 out of 483 windows...\n",
      "Processed 450 out of 483 windows...\n",
      "Total processed windows:483\n",
      "Number of reads: 350440634\n",
      "Number of valid reads: 323483858\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 323483858\n",
      "Num mapped first of pair: 0\n",
      "Num mapped second of pair: 0\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 969\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 307331884636\n",
      "referenceSize: 3101804739\n",
      "numberOfSequencedBases: 31084863167\n",
      "numberOfAs: 8673772186\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 975\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "Finished\n",
      "Processing /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D16-_10X.bam...\n",
      "Running: /Volumes/BioinforSSD/PROGRAMS_AND_ENVS/Programs/qualimap_v2.3/qualimap bamqc -bam /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D16-_10X.bam -outdir quality_check_results/D16-_10X.bam_qualimap --java-mem-size=4G\n",
      "Java memory size is set to 4G\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.3\n",
      "Built on 2023-05-19 16:57\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 35\n",
      "Max memory (Mb): 4294\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 483\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 483 windows...\n",
      "Processed 100 out of 483 windows...\n",
      "Processed 150 out of 483 windows...\n",
      "Processed 200 out of 483 windows...\n",
      "Processed 250 out of 483 windows...\n",
      "Processed 300 out of 483 windows...\n",
      "Processed 350 out of 483 windows...\n",
      "Processed 400 out of 483 windows...\n",
      "Processed 450 out of 483 windows...\n",
      "Total processed windows:483\n",
      "Number of reads: 350225800\n",
      "Number of valid reads: 329491040\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 329491040\n",
      "Num mapped first of pair: 0\n",
      "Num mapped second of pair: 0\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 793\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 245538328599\n",
      "referenceSize: 3101804739\n",
      "numberOfSequencedBases: 32307103535\n",
      "numberOfAs: 9200808446\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 800\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "Finished\n",
      "Processing /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D16+_10X.bam...\n",
      "Running: /Volumes/BioinforSSD/PROGRAMS_AND_ENVS/Programs/qualimap_v2.3/qualimap bamqc -bam /Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D16+_10X.bam -outdir quality_check_results/D16+_10X.bam_qualimap --java-mem-size=4G\n",
      "Java memory size is set to 4G\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.3\n",
      "Built on 2023-05-19 16:57\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 35\n",
      "Max memory (Mb): 4294\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 483\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 483 windows...\n",
      "Processed 100 out of 483 windows...\n",
      "Processed 150 out of 483 windows...\n",
      "Processed 200 out of 483 windows...\n",
      "Processed 250 out of 483 windows...\n",
      "Processed 300 out of 483 windows...\n",
      "Processed 350 out of 483 windows...\n",
      "Processed 400 out of 483 windows...\n",
      "Processed 450 out of 483 windows...\n",
      "Total processed windows:483\n",
      "Number of reads: 356897465\n",
      "Number of valid reads: 338026941\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 338026941\n",
      "Num mapped first of pair: 0\n",
      "Num mapped second of pair: 0\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 815\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 245361222659\n",
      "referenceSize: 3101804739\n",
      "numberOfSequencedBases: 33080233550\n",
      "numberOfAs: 9347563507\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 822\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "Finished\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[91m///\u001b[0m \u001b]8;id=106159;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2mv1.23\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /Users/noecrespo/Desktop/quality_check_results\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m282/282\u001b[0m  [0m0m  \n",
      "\u001b[?25h\u001b[34m          qualimap\u001b[0m | Found 6 BamQC reports\n",
      "\u001b[34m          samtools\u001b[0m | Found 6 flagstat reports\n",
      "\u001b[34m     write_results\u001b[0m | Data        : quality_check_results/multiqc/multiqc_data\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Quality check complete. Results are in the 'quality_check_results' directory.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[34m     write_results\u001b[0m | Report      : quality_check_results/multiqc/multiqc_report.html\n",
      "\u001b[34m           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "# Exploring quality of the data in the bam files of GSE118258\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "# List of BAM files\n",
    "bam_files = [\n",
    "    \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D0_10X.bam\",\n",
    "    \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D2_10X.bam\",\n",
    "    \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D8_10X.bam\",\n",
    "    \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D12_10X.bam\",\n",
    "    \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D16-_10X.bam\",\n",
    "    \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/D16+_10X.bam\"\n",
    "]\n",
    "\n",
    "# Full path to Qualimap\n",
    "qualimap_path = \"/Volumes/BioinforSSD/PROGRAMS_AND_ENVS/Programs/qualimap_v2.3/qualimap\"\n",
    "\n",
    "# Create output directory\n",
    "os.makedirs(\"quality_check_results\", exist_ok=True)\n",
    "\n",
    "# Function to run samtools flagstat\n",
    "def run_samtools_flagstat(bam_file):\n",
    "    output_file = f\"quality_check_results/{os.path.basename(bam_file)}_flagstat.txt\"\n",
    "    subprocess.run(f\"samtools flagstat {bam_file} > {output_file}\", shell=True)\n",
    "\n",
    "# Function to run Qualimap\n",
    "def run_qualimap(bam_file):\n",
    "    output_dir = f\"quality_check_results/{os.path.basename(bam_file)}_qualimap\"\n",
    "    command = f\"{qualimap_path} bamqc -bam {bam_file} -outdir {output_dir} --java-mem-size=4G\"\n",
    "    print(f\"Running: {command}\")\n",
    "    subprocess.run(command, shell=True)\n",
    "\n",
    "# Run quality checks on each BAM file\n",
    "for bam_file in bam_files:\n",
    "    print(f\"Processing {bam_file}...\")\n",
    "    run_samtools_flagstat(bam_file)\n",
    "    run_qualimap(bam_file)\n",
    "\n",
    "# Run MultiQC to aggregate results\n",
    "subprocess.run(\"multiqc quality_check_results -o quality_check_results/multiqc\", shell=True)\n",
    "\n",
    "print(\"Quality check complete. Results are in the 'quality_check_results' directory.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 5 rows of UMI file:\n",
      "                 AAACCTGAGGAGTTGC-1  AAACCTGAGTACTTGC-1  AAACCTGAGTATGACA-1  \\\n",
      "ENSG00000243485                   0                   0                   0   \n",
      "ENSG00000237613                   0                   0                   0   \n",
      "ENSG00000186092                   0                   0                   0   \n",
      "ENSG00000238009                   0                   0                   0   \n",
      "ENSG00000239945                   0                   0                   0   \n",
      "\n",
      "                 AAACCTGCACCATGTA-1  AAACCTGCACCTTGTC-1  AAACCTGCACGGTGTC-1  \\\n",
      "ENSG00000243485                   0                   0                   0   \n",
      "ENSG00000237613                   0                   0                   0   \n",
      "ENSG00000186092                   0                   0                   0   \n",
      "ENSG00000238009                   0                   0                   0   \n",
      "ENSG00000239945                   0                   0                   0   \n",
      "\n",
      "                 AAACCTGCAGGACGTA-1  AAACCTGCATTGGTAC-1  AAACCTGTCAAAGACA-1  \\\n",
      "ENSG00000243485                   0                   0                   0   \n",
      "ENSG00000237613                   0                   0                   0   \n",
      "ENSG00000186092                   0                   0                   0   \n",
      "ENSG00000238009                   0                   0                   0   \n",
      "ENSG00000239945                   0                   0                   0   \n",
      "\n",
      "                 AAACCTGTCAGTGCAT-1  ...  TTTCCTCTCGCACTCT-6  \\\n",
      "ENSG00000243485                   0  ...                   0   \n",
      "ENSG00000237613                   0  ...                   0   \n",
      "ENSG00000186092                   0  ...                   0   \n",
      "ENSG00000238009                   0  ...                   0   \n",
      "ENSG00000239945                   0  ...                   0   \n",
      "\n",
      "                 TTTCCTCTCTCCTATA-6  TTTGCGCCACCAGTTA-6  TTTGCGCGTCTGATTG-6  \\\n",
      "ENSG00000243485                   0                   0                   0   \n",
      "ENSG00000237613                   0                   0                   0   \n",
      "ENSG00000186092                   0                   0                   0   \n",
      "ENSG00000238009                   0                   0                   0   \n",
      "ENSG00000239945                   0                   0                   0   \n",
      "\n",
      "                 TTTGGTTCAATAACGA-6  TTTGGTTGTTACGACT-6  TTTGGTTTCTGTCTCG-6  \\\n",
      "ENSG00000243485                   0                   0                   0   \n",
      "ENSG00000237613                   0                   0                   0   \n",
      "ENSG00000186092                   0                   0                   0   \n",
      "ENSG00000238009                   0                   0                   0   \n",
      "ENSG00000239945                   0                   0                   0   \n",
      "\n",
      "                 TTTGGTTTCTTCGGTC-6  TTTGTCACAATGTAAG-6  TTTGTCATCCAAGTAC-6  \n",
      "ENSG00000243485                   0                   0                   0  \n",
      "ENSG00000237613                   0                   0                   0  \n",
      "ENSG00000186092                   0                   0                   0  \n",
      "ENSG00000238009                   0                   0                   0  \n",
      "ENSG00000239945                   0                   0                   0  \n",
      "\n",
      "[5 rows x 32138 columns]\n",
      "\n",
      "Columns:\n",
      "Index(['AAACCTGAGGAGTTGC-1', 'AAACCTGAGTACTTGC-1', 'AAACCTGAGTATGACA-1',\n",
      "       'AAACCTGCACCATGTA-1', 'AAACCTGCACCTTGTC-1', 'AAACCTGCACGGTGTC-1',\n",
      "       'AAACCTGCAGGACGTA-1', 'AAACCTGCATTGGTAC-1', 'AAACCTGTCAAAGACA-1',\n",
      "       'AAACCTGTCAGTGCAT-1',\n",
      "       ...\n",
      "       'TTTCCTCTCGCACTCT-6', 'TTTCCTCTCTCCTATA-6', 'TTTGCGCCACCAGTTA-6',\n",
      "       'TTTGCGCGTCTGATTG-6', 'TTTGGTTCAATAACGA-6', 'TTTGGTTGTTACGACT-6',\n",
      "       'TTTGGTTTCTGTCTCG-6', 'TTTGGTTTCTTCGGTC-6', 'TTTGTCACAATGTAAG-6',\n",
      "       'TTTGTCATCCAAGTAC-6'],\n",
      "      dtype='object', length=32138)\n",
      "\n",
      "Sample of index (cell barcodes):\n",
      "Index(['ENSG00000243485', 'ENSG00000237613', 'ENSG00000186092',\n",
      "       'ENSG00000238009', 'ENSG00000239945'],\n",
      "      dtype='object')\n",
      "\n",
      "Total number of lines in the file: 32739\n",
      "\n",
      "Last 5 rows of UMI file:\n",
      "                 0  0.1  0.2  0.3  1  0.4  0.5  0.6  1.1  0.7  ...  1.3849  \\\n",
      "ENSG00000215699                                                ...           \n",
      "ENSG00000215635  0    0    0    0  0    0    0    0    0    0  ...       0   \n",
      "ENSG00000268590  0    0    0    0  0    0    0    0    0    0  ...       0   \n",
      "ENSG00000251180  0    0    0    0  0    0    0    0    0    0  ...       0   \n",
      "ENSG00000215616  0    0    0    0  0    0    0    0    0    0  ...       0   \n",
      "ENSG00000215611  0    0    0    0  0    0    0    0    0    0  ...       0   \n",
      "\n",
      "                 0.27822  0.27823  1.3850  1.3851  1.3852  1.3853  0.27824  \\\n",
      "ENSG00000215699                                                              \n",
      "ENSG00000215635        0        0       0       0       0       0        0   \n",
      "ENSG00000268590        0        0       0       0       0       0        0   \n",
      "ENSG00000251180        0        0       0       0       0       0        0   \n",
      "ENSG00000215616        0        0       0       0       0       0        0   \n",
      "ENSG00000215611        0        0       0       0       0       0        0   \n",
      "\n",
      "                 0.27825  1.3854  \n",
      "ENSG00000215699                   \n",
      "ENSG00000215635        0       0  \n",
      "ENSG00000268590        0       0  \n",
      "ENSG00000251180        0       0  \n",
      "ENSG00000215616        0       0  \n",
      "ENSG00000215611        0       0  \n",
      "\n",
      "[5 rows x 32138 columns]\n"
     ]
    }
   ],
   "source": [
    "# Exploring UMI file\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "umi_file = \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/metadata/GSE118258_UMI.csv\"\n",
    "\n",
    "# Read only the first few rows\n",
    "umi_sample = pd.read_csv(umi_file, index_col=0, nrows=5)\n",
    "\n",
    "print(\"First 5 rows of UMI file:\")\n",
    "print(umi_sample)\n",
    "\n",
    "print(\"\\nColumns:\")\n",
    "print(umi_sample.columns)\n",
    "\n",
    "print(\"\\nSample of index (cell barcodes):\")\n",
    "print(umi_sample.index)\n",
    "\n",
    "# Get file info without reading the entire file\n",
    "with open(umi_file, 'r') as f:\n",
    "    num_lines = sum(1 for line in f)\n",
    "\n",
    "print(f\"\\nTotal number of lines in the file: {num_lines}\")\n",
    "\n",
    "# Read the last few lines\n",
    "tail = pd.read_csv(umi_file, index_col=0, nrows=5, skiprows=num_lines-6)\n",
    "\n",
    "print(\"\\nLast 5 rows of UMI file:\")\n",
    "print(tail)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading annotation file...\n",
      "\n",
      "Annotation file shape: (32138, 1)\n",
      "\n",
      "Columns in the annotation file:\n",
      "\n",
      "Column: Time-point\n",
      "Unique values:\n",
      "  D0: 7079\n",
      "  D2: 4988\n",
      "  D8: 3619\n",
      "  D12: 7013\n",
      "  D16_negative: 6781\n",
      "  D16_positive: 2658\n",
      "\n",
      "First few rows of the annotation file:\n",
      "                   Time-point\n",
      "Cell ID                      \n",
      "AAACCTGAGGAGTTGC-1         D0\n",
      "AAACCTGAGTACTTGC-1         D0\n",
      "AAACCTGAGTATGACA-1         D0\n",
      "AAACCTGCACCATGTA-1         D0\n",
      "AAACCTGCACCTTGTC-1         D0\n",
      "\n",
      "Last few rows of the annotation file:\n",
      "                      Time-point\n",
      "Cell ID                         \n",
      "TTTGGTTGTTACGACT-6  D16_positive\n",
      "TTTGGTTTCTGTCTCG-6  D16_positive\n",
      "TTTGGTTTCTTCGGTC-6  D16_positive\n",
      "TTTGTCACAATGTAAG-6  D16_positive\n",
      "TTTGTCATCCAAGTAC-6  D16_positive\n",
      "\n",
      "Missing values in each column:\n",
      "Time-point    0\n",
      "dtype: int64\n",
      "\n",
      "Data types of each column:\n",
      "Time-point    object\n",
      "dtype: object\n",
      "\n",
      "Sample of 5 random rows:\n",
      "                      Time-point\n",
      "Cell ID                         \n",
      "GCTGGGTTCAGCACAT-1            D0\n",
      "GGCGACTCAGCCTGTG-5  D16_negative\n",
      "GATCGATGTGACTACT-1            D0\n",
      "CTCACACTCAACACGT-5  D16_negative\n",
      "TGGACGCGTGCCTGCA-4           D12\n"
     ]
    }
   ],
   "source": [
    "# Exploring annotation file\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# File path\n",
    "annotation_file = \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/metadata/GSE118258_Annotation.txt\"\n",
    "\n",
    "# Read the annotation file\n",
    "print(\"Reading annotation file...\")\n",
    "annotation_df = pd.read_csv(annotation_file, sep='\\t', index_col=0)\n",
    "\n",
    "print(\"\\nAnnotation file shape:\", annotation_df.shape)\n",
    "\n",
    "print(\"\\nColumns in the annotation file:\")\n",
    "for column in annotation_df.columns:\n",
    "    print(f\"\\nColumn: {column}\")\n",
    "    print(\"Unique values:\")\n",
    "    unique_values = annotation_df[column].unique()\n",
    "    for value in unique_values:\n",
    "        count = annotation_df[column].value_counts()[value]\n",
    "        print(f\"  {value}: {count}\")\n",
    "\n",
    "print(\"\\nFirst few rows of the annotation file:\")\n",
    "print(annotation_df.head())\n",
    "\n",
    "print(\"\\nLast few rows of the annotation file:\")\n",
    "print(annotation_df.tail())\n",
    "\n",
    "# Check for any missing values\n",
    "print(\"\\nMissing values in each column:\")\n",
    "print(annotation_df.isnull().sum())\n",
    "\n",
    "# Check data types of each column\n",
    "print(\"\\nData types of each column:\")\n",
    "print(annotation_df.dtypes)\n",
    "\n",
    "# If there are any numeric columns, show basic statistics\n",
    "numeric_columns = annotation_df.select_dtypes(include=[np.number]).columns\n",
    "if len(numeric_columns) > 0:\n",
    "    print(\"\\nBasic statistics for numeric columns:\")\n",
    "    print(annotation_df[numeric_columns].describe())\n",
    "\n",
    "print(\"\\nSample of 5 random rows:\")\n",
    "print(annotation_df.sample(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading UMI file...\n",
      "Adding sample information...\n",
      "Adding metadata...\n",
      "Calculating QC metrics...\n",
      "Adding layers for scGPT compatibility...\n",
      "Saving h5ad file...\n",
      "H5AD file created successfully!\n",
      "\n",
      "Number of cells: 32138\n",
      "Number of genes: 32738\n",
      "Sample distribution:\n",
      "sample\n",
      "1    7079\n",
      "4    7013\n",
      "5    6781\n",
      "2    4988\n",
      "3    3619\n",
      "6    2658\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Warning: The following QC metrics are missing: n_genes\n",
      "Calculating basic metrics manually...\n",
      "\n",
      "Mean genes per cell: 4007.88\n",
      "Median genes per cell: 3986.00\n",
      "Mean counts per cell: 4867.19\n",
      "Median counts per cell: 4743.17\n",
      "\n",
      "Metadata columns: ['sample', 'Time-point', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'n_genes']\n",
      "\n",
      "AnnData object information:\n",
      "AnnData object with n_obs × n_vars = 32138 × 32738\n",
      "    obs: 'sample', 'Time-point', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'n_genes'\n",
      "    var: 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
      "    layers: 'counts', 'log1p'\n",
      "\n",
      "First few rows of adata.obs:\n",
      "                   sample Time-point  n_genes_by_counts  \\\n",
      "AAACCTGAGGAGTTGC-1      1         D0               4141   \n",
      "AAACCTGAGTACTTGC-1      1         D0               4568   \n",
      "AAACCTGAGTATGACA-1      1         D0               4343   \n",
      "AAACCTGCACCATGTA-1      1         D0               4056   \n",
      "AAACCTGCACCTTGTC-1      1         D0               3917   \n",
      "\n",
      "                    log1p_n_genes_by_counts  total_counts  log1p_total_counts  \\\n",
      "AAACCTGAGGAGTTGC-1                 8.328934   4971.494629            9.857025   \n",
      "AAACCTGAGTACTTGC-1                 8.427050   5732.989258           10.062029   \n",
      "AAACCTGAGTATGACA-1                 8.376551   5557.826172           10.312911   \n",
      "AAACCTGCACCATGTA-1                 8.308199   4997.662109            9.952087   \n",
      "AAACCTGCACCTTGTC-1                 8.273337   4655.264648            9.746132   \n",
      "\n",
      "                    pct_counts_in_top_50_genes  pct_counts_in_top_100_genes  \\\n",
      "AAACCTGAGGAGTTGC-1                   26.357970                    38.510293   \n",
      "AAACCTGAGTACTTGC-1                   25.180286                    36.859398   \n",
      "AAACCTGAGTATGACA-1                   36.811873                    51.029285   \n",
      "AAACCTGCACCATGTA-1                   28.335318                    41.019290   \n",
      "AAACCTGCACCTTGTC-1                   26.365073                    38.719494   \n",
      "\n",
      "                    pct_counts_in_top_200_genes  pct_counts_in_top_500_genes  \\\n",
      "AAACCTGAGGAGTTGC-1                    50.316903                    64.595883   \n",
      "AAACCTGAGTACTTGC-1                    48.645189                    63.383828   \n",
      "AAACCTGAGTATGACA-1                    61.292250                    73.122385   \n",
      "AAACCTGCACCATGTA-1                    52.993570                    67.263634   \n",
      "AAACCTGCACCTTGTC-1                    49.768830                    63.884825   \n",
      "\n",
      "                    n_genes  \n",
      "AAACCTGAGGAGTTGC-1     4141  \n",
      "AAACCTGAGTACTTGC-1     4568  \n",
      "AAACCTGAGTATGACA-1     4343  \n",
      "AAACCTGCACCATGTA-1     4056  \n",
      "AAACCTGCACCTTGTC-1     3917  \n",
      "\n",
      "First few rows of adata.var:\n",
      "                 n_cells_by_counts  mean_counts  log1p_mean_counts  \\\n",
      "ENSG00000243485                  0     0.000000           0.000000   \n",
      "ENSG00000237613                  0     0.000000           0.000000   \n",
      "ENSG00000186092                  0     0.000000           0.000000   \n",
      "ENSG00000238009                139     0.004387           0.004378   \n",
      "ENSG00000239945                  0     0.000000           0.000000   \n",
      "\n",
      "                 pct_dropout_by_counts  total_counts  log1p_total_counts  \n",
      "ENSG00000243485              100.00000           0.0            0.000000  \n",
      "ENSG00000237613              100.00000           0.0            0.000000  \n",
      "ENSG00000186092              100.00000           0.0            0.000000  \n",
      "ENSG00000238009               99.56749         141.0            4.955827  \n",
      "ENSG00000239945              100.00000           0.0            0.000000  \n"
     ]
    }
   ],
   "source": [
    "# Creating h5ad file\n",
    "\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "\n",
    "# File paths\n",
    "umi_file = \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/metadata/GSE118258_UMI.csv\"\n",
    "annotation_file = \"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/metadata/GSE118258_Annotation.txt\"\n",
    "\n",
    "print(\"Reading UMI file...\")\n",
    "adata = sc.read_csv(umi_file, first_column_names=True).T\n",
    "\n",
    "print(\"Adding sample information...\")\n",
    "adata.obs['sample'] = [cell.split('-')[-1] for cell in adata.obs_names]\n",
    "\n",
    "print(\"Adding metadata...\")\n",
    "annotation_df = pd.read_csv(annotation_file, sep='\\t', index_col=0)\n",
    "adata.obs = adata.obs.join(annotation_df, how='left')\n",
    "\n",
    "print(\"Calculating QC metrics...\")\n",
    "sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "\n",
    "print(\"Adding layers for scGPT compatibility...\")\n",
    "adata.layers[\"counts\"] = adata.X.copy()\n",
    "adata.layers[\"log1p\"] = sc.pp.log1p(adata.X)\n",
    "\n",
    "print(\"Saving h5ad file...\")\n",
    "adata.write_h5ad(\"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/GSE118258_scRNA_seq.h5ad\")\n",
    "\n",
    "print(\"H5AD file created successfully!\")\n",
    "\n",
    "# Print some basic statistics\n",
    "print(f\"\\nNumber of cells: {adata.n_obs}\")\n",
    "print(f\"Number of genes: {adata.n_vars}\")\n",
    "print(\"Sample distribution:\")\n",
    "print(adata.obs['sample'].value_counts())\n",
    "\n",
    "# Check if QC metrics are present\n",
    "qc_metrics = ['n_genes_by_counts', 'total_counts', 'n_genes']\n",
    "missing_metrics = [metric for metric in qc_metrics if metric not in adata.obs.columns]\n",
    "\n",
    "if missing_metrics:\n",
    "    print(f\"\\nWarning: The following QC metrics are missing: {', '.join(missing_metrics)}\")\n",
    "    print(\"Calculating basic metrics manually...\")\n",
    "    \n",
    "    adata.obs['n_genes'] = (adata.X > 0).sum(axis=1)\n",
    "    adata.obs['total_counts'] = adata.X.sum(axis=1)\n",
    "\n",
    "print(f\"\\nMean genes per cell: {adata.obs['n_genes'].mean():.2f}\")\n",
    "print(f\"Median genes per cell: {adata.obs['n_genes'].median():.2f}\")\n",
    "print(f\"Mean counts per cell: {adata.obs['total_counts'].mean():.2f}\")\n",
    "print(f\"Median counts per cell: {adata.obs['total_counts'].median():.2f}\")\n",
    "\n",
    "print(\"\\nMetadata columns:\", adata.obs.columns.tolist())\n",
    "\n",
    "# Additional information about the AnnData object\n",
    "print(\"\\nAnnData object information:\")\n",
    "print(adata)\n",
    "\n",
    "# Check the contents of adata.obs\n",
    "print(\"\\nFirst few rows of adata.obs:\")\n",
    "print(adata.obs.head())\n",
    "\n",
    "# Check the contents of adata.var\n",
    "print(\"\\nFirst few rows of adata.var:\")\n",
    "print(adata.var.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object information:\n",
      "AnnData object with n_obs × n_vars = 32138 × 32738\n",
      "    obs: 'Time-point', 'sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'leiden'\n",
      "    var: 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
      "    uns: 'hvg', 'leiden', 'log1p', 'neighbors', 'pca', 'umap'\n",
      "    obsm: 'X_pca', 'X_umap'\n",
      "    varm: 'PCs'\n",
      "    layers: 'counts', 'log1p'\n",
      "    obsp: 'connectivities', 'distances'\n",
      "\n",
      "Shape of adata: (32138, 32738)\n",
      "\n",
      "Columns in adata.obs:\n",
      "Index(['Time-point', 'sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts',\n",
      "       'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes',\n",
      "       'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes',\n",
      "       'pct_counts_in_top_500_genes', 'leiden'],\n",
      "      dtype='object')\n",
      "\n",
      "First few rows of adata.obs:\n",
      "                   Time-point sample  n_genes_by_counts  \\\n",
      "AAACCTGAGGAGTTGC-1         D0      1               4141   \n",
      "AAACCTGAGTACTTGC-1         D0      1               4568   \n",
      "AAACCTGAGTATGACA-1         D0      1               4343   \n",
      "AAACCTGCACCATGTA-1         D0      1               4056   \n",
      "AAACCTGCACCTTGTC-1         D0      1               3917   \n",
      "\n",
      "                    log1p_n_genes_by_counts  total_counts  log1p_total_counts  \\\n",
      "AAACCTGAGGAGTTGC-1                 8.328934       19091.0            9.857025   \n",
      "AAACCTGAGTACTTGC-1                 8.427050       23435.0           10.062029   \n",
      "AAACCTGAGTATGACA-1                 8.376551       30118.0           10.312911   \n",
      "AAACCTGCACCATGTA-1                 8.308199       20995.0            9.952087   \n",
      "AAACCTGCACCTTGTC-1                 8.273337       17087.0            9.746132   \n",
      "\n",
      "                    pct_counts_in_top_50_genes  pct_counts_in_top_100_genes  \\\n",
      "AAACCTGAGGAGTTGC-1                   26.357970                    38.510293   \n",
      "AAACCTGAGTACTTGC-1                   25.180286                    36.859398   \n",
      "AAACCTGAGTATGACA-1                   36.811873                    51.029285   \n",
      "AAACCTGCACCATGTA-1                   28.335318                    41.019290   \n",
      "AAACCTGCACCTTGTC-1                   26.365073                    38.719494   \n",
      "\n",
      "                    pct_counts_in_top_200_genes  pct_counts_in_top_500_genes  \\\n",
      "AAACCTGAGGAGTTGC-1                    50.316903                    64.595883   \n",
      "AAACCTGAGTACTTGC-1                    48.645189                    63.383828   \n",
      "AAACCTGAGTATGACA-1                    61.292250                    73.122385   \n",
      "AAACCTGCACCATGTA-1                    52.993570                    67.263634   \n",
      "AAACCTGCACCTTGTC-1                    49.768830                    63.884825   \n",
      "\n",
      "                   leiden  \n",
      "AAACCTGAGGAGTTGC-1      5  \n",
      "AAACCTGAGTACTTGC-1     15  \n",
      "AAACCTGAGTATGACA-1      8  \n",
      "AAACCTGCACCATGTA-1      8  \n",
      "AAACCTGCACCTTGTC-1      5  \n",
      "\n",
      "Columns in adata.var:\n",
      "Index(['n_cells_by_counts', 'mean_counts', 'log1p_mean_counts',\n",
      "       'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts',\n",
      "       'highly_variable', 'means', 'dispersions', 'dispersions_norm'],\n",
      "      dtype='object')\n",
      "\n",
      "First few rows of adata.var:\n",
      "                 n_cells_by_counts  mean_counts  log1p_mean_counts  \\\n",
      "ENSG00000243485                  0     0.000000           0.000000   \n",
      "ENSG00000237613                  0     0.000000           0.000000   \n",
      "ENSG00000186092                  0     0.000000           0.000000   \n",
      "ENSG00000238009                139     0.004387           0.004378   \n",
      "ENSG00000239945                  0     0.000000           0.000000   \n",
      "\n",
      "                 pct_dropout_by_counts  total_counts  log1p_total_counts  \\\n",
      "ENSG00000243485              100.00000           0.0            0.000000   \n",
      "ENSG00000237613              100.00000           0.0            0.000000   \n",
      "ENSG00000186092              100.00000           0.0            0.000000   \n",
      "ENSG00000238009               99.56749         141.0            4.955827   \n",
      "ENSG00000239945              100.00000           0.0            0.000000   \n",
      "\n",
      "                 highly_variable         means  dispersions  dispersions_norm  \n",
      "ENSG00000243485            False  1.000000e-12          NaN               NaN  \n",
      "ENSG00000237613            False  1.000000e-12          NaN               NaN  \n",
      "ENSG00000186092            False  1.000000e-12          NaN               NaN  \n",
      "ENSG00000238009            False  1.837276e-03    -0.547374         -0.100593  \n",
      "ENSG00000239945            False  1.000000e-12          NaN               NaN  \n",
      "\n",
      "Layers in adata:\n",
      "['counts', 'log1p']\n",
      "\n",
      "Sample of unique values in adata.obs:\n",
      "\n",
      "Time-point:\n",
      "['D0', 'D2', 'D8', 'D12', 'D16_negative', 'D16_positive']\n",
      "Total unique values: 6\n",
      "\n",
      "sample:\n",
      "['1', '2', '3', '4', '5', '6']\n",
      "Total unique values: 6\n",
      "\n",
      "n_genes_by_counts:\n",
      "Min: 293, Max: 8746\n",
      "Total unique values: 4634\n",
      "\n",
      "log1p_n_genes_by_counts:\n",
      "Min: 5.683579767338681, Max: 9.076466063419854\n",
      "Total unique values: 4634\n",
      "\n",
      "total_counts:\n",
      "Min: 4445.0, Max: 161511.0\n",
      "Total unique values: 20480\n",
      "\n",
      "log1p_total_counts:\n",
      "Min: 8.399760246276855, Max: 11.992334365844727\n",
      "Total unique values: 20480\n",
      "\n",
      "pct_counts_in_top_50_genes:\n",
      "Min: 19.722428748451055, Max: 95.23809523809523\n",
      "Total unique values: 32095\n",
      "\n",
      "pct_counts_in_top_100_genes:\n",
      "Min: 29.390334572490705, Max: 96.21791103272585\n",
      "Total unique values: 32095\n",
      "\n",
      "pct_counts_in_top_200_genes:\n",
      "Min: 39.935563816604706, Max: 98.17754262198707\n",
      "Total unique values: 32097\n",
      "\n",
      "pct_counts_in_top_500_genes:\n",
      "Min: 54.62070749531139, Max: 100.0\n",
      "Total unique values: 32086\n",
      "\n",
      "leiden:\n",
      "['5', '15', '8', '4', '0', '12', '14', '11', '7', '3', '...']\n",
      "Total unique values: 22\n",
      "\n",
      "Sample of unique values in adata.var:\n",
      "\n",
      "n_cells_by_counts:\n",
      "Min: 0, Max: 32138\n",
      "Total unique values: 9956\n",
      "\n",
      "mean_counts:\n",
      "Min: 0.0, Max: 299.4154357910156\n",
      "Total unique values: 10993\n",
      "\n",
      "log1p_mean_counts:\n",
      "Min: 0.0, Max: 5.705166339874268\n",
      "Total unique values: 10993\n",
      "\n",
      "pct_dropout_by_counts:\n",
      "Min: 0.0, Max: 100.0\n",
      "Total unique values: 9956\n",
      "\n",
      "total_counts:\n",
      "Min: 0.0, Max: 9622613.0\n",
      "Total unique values: 10993\n",
      "\n",
      "log1p_total_counts:\n",
      "Min: 0.0, Max: 16.079626083374023\n",
      "Total unique values: 10993\n",
      "\n",
      "highly_variable:\n",
      "[np.False_, np.True_]\n",
      "Total unique values: 2\n",
      "\n",
      "means:\n",
      "Min: 9.999999999995e-13, Max: 4.878298304589947\n",
      "Total unique values: 26258\n",
      "\n",
      "dispersions:\n",
      "Min: nan, Max: nan\n",
      "Total unique values: 26258\n",
      "\n",
      "dispersions_norm:\n",
      "Min: nan, Max: nan\n",
      "Total unique values: 26255\n"
     ]
    }
   ],
   "source": [
    "# Lets explore again the h5ad file created:\n",
    "\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Load the data\n",
    "adata = sc.read_h5ad(\"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/H5AD_FILES/GSE118258_scRNA_seq_analyzed.h5ad\")\n",
    "\n",
    "print(\"AnnData object information:\")\n",
    "print(adata)\n",
    "\n",
    "print(\"\\nShape of adata:\", adata.shape)\n",
    "\n",
    "print(\"\\nColumns in adata.obs:\")\n",
    "print(adata.obs.columns)\n",
    "\n",
    "print(\"\\nFirst few rows of adata.obs:\")\n",
    "print(adata.obs.head())\n",
    "\n",
    "print(\"\\nColumns in adata.var:\")\n",
    "print(adata.var.columns)\n",
    "\n",
    "print(\"\\nFirst few rows of adata.var:\")\n",
    "print(adata.var.head())\n",
    "\n",
    "print(\"\\nLayers in adata:\")\n",
    "print(list(adata.layers.keys()))\n",
    "\n",
    "# Function to print unique values\n",
    "def print_unique_values(df, name, max_values=10):\n",
    "    print(f\"\\nSample of unique values in {name}:\")\n",
    "    for column in df.columns:\n",
    "        unique_values = df[column].unique()\n",
    "        print(f\"\\n{column}:\")\n",
    "        if df[column].dtype == 'object' or df[column].dtype.name == 'category':\n",
    "            print(list(unique_values[:max_values]) + ([\"...\"] if len(unique_values) > max_values else []))\n",
    "        elif np.issubdtype(df[column].dtype, np.number):\n",
    "            if len(unique_values) > max_values:\n",
    "                print(f\"Min: {unique_values.min()}, Max: {unique_values.max()}\")\n",
    "            else:\n",
    "                print(list(unique_values))\n",
    "        else:\n",
    "            print(list(unique_values[:max_values]) + ([\"...\"] if len(unique_values) > max_values else []))\n",
    "        print(f\"Total unique values: {len(unique_values)}\")\n",
    "\n",
    "print_unique_values(adata.obs, \"adata.obs\")\n",
    "print_unique_values(adata.var, \"adata.var\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated columns in adata.var:\n",
      "Index(['gene_name', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts',\n",
      "       'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts',\n",
      "       'highly_variable', 'means', 'dispersions', 'dispersions_norm'],\n",
      "      dtype='object')\n",
      "\n",
      "First few rows of adata.var:\n",
      "                       gene_name  n_cells_by_counts  mean_counts  \\\n",
      "ENSG00000243485      MIR1302-2HG                  0     0.000000   \n",
      "ENSG00000237613          FAM138A                  0     0.000000   \n",
      "ENSG00000186092            OR4F5                  0     0.000000   \n",
      "ENSG00000238009  ENSG00000238009                139     0.004387   \n",
      "ENSG00000239945  ENSG00000239945                  0     0.000000   \n",
      "\n",
      "                 log1p_mean_counts  pct_dropout_by_counts  total_counts  \\\n",
      "ENSG00000243485           0.000000              100.00000           0.0   \n",
      "ENSG00000237613           0.000000              100.00000           0.0   \n",
      "ENSG00000186092           0.000000              100.00000           0.0   \n",
      "ENSG00000238009           0.004378               99.56749         141.0   \n",
      "ENSG00000239945           0.000000              100.00000           0.0   \n",
      "\n",
      "                 log1p_total_counts  highly_variable         means  \\\n",
      "ENSG00000243485            0.000000            False  1.000000e-12   \n",
      "ENSG00000237613            0.000000            False  1.000000e-12   \n",
      "ENSG00000186092            0.000000            False  1.000000e-12   \n",
      "ENSG00000238009            4.955827            False  1.837276e-03   \n",
      "ENSG00000239945            0.000000            False  1.000000e-12   \n",
      "\n",
      "                 dispersions  dispersions_norm  \n",
      "ENSG00000243485          NaN               NaN  \n",
      "ENSG00000237613          NaN               NaN  \n",
      "ENSG00000186092          NaN               NaN  \n",
      "ENSG00000238009    -0.547374         -0.100593  \n",
      "ENSG00000239945          NaN               NaN  \n",
      "\n",
      "Shape of updated adata: (32138, 32738)\n",
      "\n",
      "Sample of unique values in adata.var:\n",
      "\n",
      "gene_name:\n",
      "['MIR1302-2HG', 'FAM138A', 'OR4F5', 'ENSG00000238009', 'ENSG00000239945', 'ENSG00000237683', 'ENSG00000239906', 'ENSG00000241599', 'RPL23AP21', 'ENSG00000237094', '...']\n",
      "Total unique values: 32713\n",
      "\n",
      "n_cells_by_counts:\n",
      "Min: 0, Max: 32138\n",
      "Total unique values: 9956\n",
      "\n",
      "mean_counts:\n",
      "Min: 0.0, Max: 299.4154357910156\n",
      "Total unique values: 10993\n",
      "\n",
      "log1p_mean_counts:\n",
      "Min: 0.0, Max: 5.705166339874268\n",
      "Total unique values: 10993\n",
      "\n",
      "pct_dropout_by_counts:\n",
      "Min: 0.0, Max: 100.0\n",
      "Total unique values: 9956\n",
      "\n",
      "total_counts:\n",
      "Min: 0.0, Max: 9622613.0\n",
      "Total unique values: 10993\n",
      "\n",
      "log1p_total_counts:\n",
      "Min: 0.0, Max: 16.079626083374023\n",
      "Total unique values: 10993\n",
      "\n",
      "highly_variable:\n",
      "[np.False_, np.True_]\n",
      "Total unique values: 2\n",
      "\n",
      "means:\n",
      "Min: 9.999999999995e-13, Max: 4.878298304589947\n",
      "Total unique values: 26258\n",
      "\n",
      "dispersions:\n",
      "Min: nan, Max: nan\n",
      "Total unique values: 26258\n",
      "\n",
      "dispersions_norm:\n",
      "Min: nan, Max: nan\n",
      "Total unique values: 26255\n"
     ]
    }
   ],
   "source": [
    "# Now we add gene_name info\n",
    "import scanpy as sc\n",
    "import mygene\n",
    "import pandas as pd\n",
    "\n",
    "# Load the data\n",
    "adata = sc.read_h5ad(\"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/H5AD_FILES/GSE118258_scRNA_seq_analyzed.h5ad\")\n",
    "\n",
    "# Initialize mygene client\n",
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "# Get the current gene IDs (Ensembl IDs)\n",
    "gene_ids = adata.var.index.tolist()\n",
    "\n",
    "# Query mygene for gene names\n",
    "gene_info = mg.querymany(gene_ids, scopes='ensembl.gene', fields='symbol', species='human', verbose=False)\n",
    "\n",
    "# Create a dictionary mapping gene IDs to gene names\n",
    "gene_name_dict = {info['query']: info.get('symbol', info['query']) for info in gene_info if 'query' in info}\n",
    "\n",
    "# Add gene names to adata.var\n",
    "adata.var['gene_name'] = adata.var.index.map(gene_name_dict)\n",
    "\n",
    "# Ensure 'gene_name' is filled for all genes\n",
    "adata.var.loc[adata.var['gene_name'].isnull(), 'gene_name'] = adata.var.index[adata.var['gene_name'].isnull()]\n",
    "\n",
    "# Move 'gene_name' to be the first column\n",
    "cols = adata.var.columns.tolist()\n",
    "cols = ['gene_name'] + [col for col in cols if col != 'gene_name']\n",
    "adata.var = adata.var[cols]\n",
    "\n",
    "# Save the updated AnnData object\n",
    "adata.write_h5ad(\"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/H5AD_FILES/GSE118258_scRNA_seq_analyzed_with_gene_names.h5ad\")\n",
    "\n",
    "# Print some information about the updated object\n",
    "print(\"Updated columns in adata.var:\")\n",
    "print(adata.var.columns)\n",
    "\n",
    "print(\"\\nFirst few rows of adata.var:\")\n",
    "print(adata.var.head())\n",
    "\n",
    "print(\"\\nShape of updated adata:\", adata.shape)\n",
    "\n",
    "# Function to print unique values\n",
    "def print_unique_values(df, name, max_values=10):\n",
    "    print(f\"\\nSample of unique values in {name}:\")\n",
    "    for column in df.columns:\n",
    "        unique_values = df[column].unique()\n",
    "        print(f\"\\n{column}:\")\n",
    "        if df[column].dtype == 'object' or df[column].dtype.name == 'category':\n",
    "            print(list(unique_values[:max_values]) + ([\"...\"] if len(unique_values) > max_values else []))\n",
    "        elif pd.api.types.is_numeric_dtype(df[column]):\n",
    "            if len(unique_values) > max_values:\n",
    "                print(f\"Min: {unique_values.min()}, Max: {unique_values.max()}\")\n",
    "            else:\n",
    "                print(list(unique_values))\n",
    "        else:\n",
    "            print(list(unique_values[:max_values]) + ([\"...\"] if len(unique_values) > max_values else []))\n",
    "        print(f\"Total unique values: {len(unique_values)}\")\n",
    "\n",
    "# Print unique values in adata.var\n",
    "print_unique_values(adata.var, \"adata.var\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object information:\n",
      "AnnData object with n_obs × n_vars = 32138 × 32738\n",
      "    obs: 'Time-point', 'sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'leiden'\n",
      "    var: 'gene_name', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
      "    uns: 'hvg', 'leiden', 'log1p', 'neighbors', 'pca', 'umap'\n",
      "    obsm: 'X_pca', 'X_umap'\n",
      "    varm: 'PCs'\n",
      "    layers: 'counts', 'log1p'\n",
      "    obsp: 'connectivities', 'distances'\n",
      "\n",
      "Shape of adata: (32138, 32738)\n",
      "\n",
      "Columns in adata.obs:\n",
      "Index(['Time-point', 'sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts',\n",
      "       'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes',\n",
      "       'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes',\n",
      "       'pct_counts_in_top_500_genes', 'leiden'],\n",
      "      dtype='object')\n",
      "\n",
      "First few rows of adata.obs:\n",
      "                   Time-point sample  n_genes_by_counts  \\\n",
      "AAACCTGAGGAGTTGC-1         D0      1               4141   \n",
      "AAACCTGAGTACTTGC-1         D0      1               4568   \n",
      "AAACCTGAGTATGACA-1         D0      1               4343   \n",
      "AAACCTGCACCATGTA-1         D0      1               4056   \n",
      "AAACCTGCACCTTGTC-1         D0      1               3917   \n",
      "\n",
      "                    log1p_n_genes_by_counts  total_counts  log1p_total_counts  \\\n",
      "AAACCTGAGGAGTTGC-1                 8.328934       19091.0            9.857025   \n",
      "AAACCTGAGTACTTGC-1                 8.427050       23435.0           10.062029   \n",
      "AAACCTGAGTATGACA-1                 8.376551       30118.0           10.312911   \n",
      "AAACCTGCACCATGTA-1                 8.308199       20995.0            9.952087   \n",
      "AAACCTGCACCTTGTC-1                 8.273337       17087.0            9.746132   \n",
      "\n",
      "                    pct_counts_in_top_50_genes  pct_counts_in_top_100_genes  \\\n",
      "AAACCTGAGGAGTTGC-1                   26.357970                    38.510293   \n",
      "AAACCTGAGTACTTGC-1                   25.180286                    36.859398   \n",
      "AAACCTGAGTATGACA-1                   36.811873                    51.029285   \n",
      "AAACCTGCACCATGTA-1                   28.335318                    41.019290   \n",
      "AAACCTGCACCTTGTC-1                   26.365073                    38.719494   \n",
      "\n",
      "                    pct_counts_in_top_200_genes  pct_counts_in_top_500_genes  \\\n",
      "AAACCTGAGGAGTTGC-1                    50.316903                    64.595883   \n",
      "AAACCTGAGTACTTGC-1                    48.645189                    63.383828   \n",
      "AAACCTGAGTATGACA-1                    61.292250                    73.122385   \n",
      "AAACCTGCACCATGTA-1                    52.993570                    67.263634   \n",
      "AAACCTGCACCTTGTC-1                    49.768830                    63.884825   \n",
      "\n",
      "                   leiden  \n",
      "AAACCTGAGGAGTTGC-1      5  \n",
      "AAACCTGAGTACTTGC-1     15  \n",
      "AAACCTGAGTATGACA-1      8  \n",
      "AAACCTGCACCATGTA-1      8  \n",
      "AAACCTGCACCTTGTC-1      5  \n",
      "\n",
      "Columns in adata.var:\n",
      "Index(['gene_name', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts',\n",
      "       'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts',\n",
      "       'highly_variable', 'means', 'dispersions', 'dispersions_norm'],\n",
      "      dtype='object')\n",
      "\n",
      "First few rows of adata.var:\n",
      "                       gene_name  n_cells_by_counts  mean_counts  \\\n",
      "ENSG00000243485      MIR1302-2HG                  0     0.000000   \n",
      "ENSG00000237613          FAM138A                  0     0.000000   \n",
      "ENSG00000186092            OR4F5                  0     0.000000   \n",
      "ENSG00000238009  ENSG00000238009                139     0.004387   \n",
      "ENSG00000239945  ENSG00000239945                  0     0.000000   \n",
      "\n",
      "                 log1p_mean_counts  pct_dropout_by_counts  total_counts  \\\n",
      "ENSG00000243485           0.000000              100.00000           0.0   \n",
      "ENSG00000237613           0.000000              100.00000           0.0   \n",
      "ENSG00000186092           0.000000              100.00000           0.0   \n",
      "ENSG00000238009           0.004378               99.56749         141.0   \n",
      "ENSG00000239945           0.000000              100.00000           0.0   \n",
      "\n",
      "                 log1p_total_counts  highly_variable         means  \\\n",
      "ENSG00000243485            0.000000            False  1.000000e-12   \n",
      "ENSG00000237613            0.000000            False  1.000000e-12   \n",
      "ENSG00000186092            0.000000            False  1.000000e-12   \n",
      "ENSG00000238009            4.955827            False  1.837276e-03   \n",
      "ENSG00000239945            0.000000            False  1.000000e-12   \n",
      "\n",
      "                 dispersions  dispersions_norm  \n",
      "ENSG00000243485          NaN               NaN  \n",
      "ENSG00000237613          NaN               NaN  \n",
      "ENSG00000186092          NaN               NaN  \n",
      "ENSG00000238009    -0.547374         -0.100593  \n",
      "ENSG00000239945          NaN               NaN  \n",
      "\n",
      "Layers in adata:\n",
      "['counts', 'log1p']\n",
      "\n",
      "Sample of unique values in adata.obs:\n",
      "\n",
      "Time-point:\n",
      "['D0', 'D2', 'D8', 'D12', 'D16_negative', 'D16_positive']\n",
      "Total unique values: 6\n",
      "\n",
      "sample:\n",
      "['1', '2', '3', '4', '5', '6']\n",
      "Total unique values: 6\n",
      "\n",
      "n_genes_by_counts:\n",
      "Min: 293, Max: 8746\n",
      "Total unique values: 4634\n",
      "\n",
      "log1p_n_genes_by_counts:\n",
      "Min: 5.683579767338681, Max: 9.076466063419854\n",
      "Total unique values: 4634\n",
      "\n",
      "total_counts:\n",
      "Min: 4445.0, Max: 161511.0\n",
      "Total unique values: 20480\n",
      "\n",
      "log1p_total_counts:\n",
      "Min: 8.399760246276855, Max: 11.992334365844727\n",
      "Total unique values: 20480\n",
      "\n",
      "pct_counts_in_top_50_genes:\n",
      "Min: 19.722428748451055, Max: 95.23809523809523\n",
      "Total unique values: 32095\n",
      "\n",
      "pct_counts_in_top_100_genes:\n",
      "Min: 29.390334572490705, Max: 96.21791103272585\n",
      "Total unique values: 32095\n",
      "\n",
      "pct_counts_in_top_200_genes:\n",
      "Min: 39.935563816604706, Max: 98.17754262198707\n",
      "Total unique values: 32097\n",
      "\n",
      "pct_counts_in_top_500_genes:\n",
      "Min: 54.62070749531139, Max: 100.0\n",
      "Total unique values: 32086\n",
      "\n",
      "leiden:\n",
      "['5', '15', '8', '4', '0', '12', '14', '11', '7', '3', '...']\n",
      "Total unique values: 22\n",
      "\n",
      "Sample of unique values in adata.var:\n",
      "\n",
      "gene_name:\n",
      "['MIR1302-2HG', 'FAM138A', 'OR4F5', 'ENSG00000238009', 'ENSG00000239945', 'ENSG00000237683', 'ENSG00000239906', 'ENSG00000241599', 'RPL23AP21', 'ENSG00000237094', '...']\n",
      "Total unique values: 32713\n",
      "\n",
      "n_cells_by_counts:\n",
      "Min: 0, Max: 32138\n",
      "Total unique values: 9956\n",
      "\n",
      "mean_counts:\n",
      "Min: 0.0, Max: 299.4154357910156\n",
      "Total unique values: 10993\n",
      "\n",
      "log1p_mean_counts:\n",
      "Min: 0.0, Max: 5.705166339874268\n",
      "Total unique values: 10993\n",
      "\n",
      "pct_dropout_by_counts:\n",
      "Min: 0.0, Max: 100.0\n",
      "Total unique values: 9956\n",
      "\n",
      "total_counts:\n",
      "Min: 0.0, Max: 9622613.0\n",
      "Total unique values: 10993\n",
      "\n",
      "log1p_total_counts:\n",
      "Min: 0.0, Max: 16.079626083374023\n",
      "Total unique values: 10993\n",
      "\n",
      "highly_variable:\n",
      "[np.False_, np.True_]\n",
      "Total unique values: 2\n",
      "\n",
      "means:\n",
      "Min: 9.999999999995e-13, Max: 4.878298304589947\n",
      "Total unique values: 26258\n",
      "\n",
      "dispersions:\n",
      "Min: nan, Max: nan\n",
      "Total unique values: 26258\n",
      "\n",
      "dispersions_norm:\n",
      "Min: nan, Max: nan\n",
      "Total unique values: 26255\n"
     ]
    }
   ],
   "source": [
    "# Now we explore again the h5ad file after adding gene name info\n",
    "\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Load the data\n",
    "adata = sc.read_h5ad(\"/Volumes/BioinforSSD/DATASETS_JUNE/GSE118258/H5AD_FILES/GSE118258_scRNA_seq_analyzed_with_gene_names.h5ad\")\n",
    "\n",
    "print(\"AnnData object information:\")\n",
    "print(adata)\n",
    "\n",
    "print(\"\\nShape of adata:\", adata.shape)\n",
    "\n",
    "print(\"\\nColumns in adata.obs:\")\n",
    "print(adata.obs.columns)\n",
    "\n",
    "print(\"\\nFirst few rows of adata.obs:\")\n",
    "print(adata.obs.head())\n",
    "\n",
    "print(\"\\nColumns in adata.var:\")\n",
    "print(adata.var.columns)\n",
    "\n",
    "print(\"\\nFirst few rows of adata.var:\")\n",
    "print(adata.var.head())\n",
    "\n",
    "print(\"\\nLayers in adata:\")\n",
    "print(list(adata.layers.keys()))\n",
    "\n",
    "# Function to print unique values\n",
    "def print_unique_values(df, name, max_values=10):\n",
    "    print(f\"\\nSample of unique values in {name}:\")\n",
    "    for column in df.columns:\n",
    "        unique_values = df[column].unique()\n",
    "        print(f\"\\n{column}:\")\n",
    "        if df[column].dtype == 'object' or df[column].dtype.name == 'category':\n",
    "            print(list(unique_values[:max_values]) + ([\"...\"] if len(unique_values) > max_values else []))\n",
    "        elif np.issubdtype(df[column].dtype, np.number):\n",
    "            if len(unique_values) > max_values:\n",
    "                print(f\"Min: {unique_values.min()}, Max: {unique_values.max()}\")\n",
    "            else:\n",
    "                print(list(unique_values))\n",
    "        else:\n",
    "            print(list(unique_values[:max_values]) + ([\"...\"] if len(unique_values) > max_values else []))\n",
    "        print(f\"Total unique values: {len(unique_values)}\")\n",
    "\n",
    "print_unique_values(adata.obs, \"adata.obs\")\n",
    "print_unique_values(adata.var, \"adata.var\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
